domain/src/main/java/org/oppia/android/domain/translation/TranslationController.kt

```diff
@@ -164,22 +168,19 @@ class TranslationController @Inject constructor(
    *     result is ever provided). The payload of the data provider is the *previous* selection
    *     state.
    */
-  fun updateAppLanguage(profileId: ProfileId, selection: AppLanguageSelection): DataProvider<Any> {
-    val deferred = retrieveLanguageContentCacheStore(profileId).storeDataAsync(
-      updateInMemoryCache = true
-    ) {
-      it.toBuilder().apply { selectedLanguage = selection.selectedLanguage }.build()
-    }
-    return dataProviders.createInMemoryDataProviderAsync(
-      UPDATE_APP_LANGUAGE_DATA_PROVIDER_ID
-    ) {
-      updateAppLanguageSelection(profileId, selection)
-      try {
-        return@createInMemoryDataProviderAsync AsyncResult.Success(deferred.await())
-      } catch (e: Exception) {
-        return@createInMemoryDataProviderAsync AsyncResult.Failure(e)
-      }
+  fun updateAppLanguage(
+    profileId: ProfileId,
+    selection: AppLanguageSelection
+  ): DataProvider<AppLanguageSelection> {
+    val cacheStore = retrieveLanguageContentCacheStore(profileId)
+    cacheStore.storeDataAsync(updateInMemoryCache = true) {
+      selection
+    }.invokeOnCompletion { exception ->
+      oppiaLogger.e(
+        "TranslationController", "Failed updating language: $exception"
+      )
     }
+    return cacheStore
```

Comment:
Why return the same cache store? It could just be retrieved from getAppLanguageSelection above. I'd actually expect this function to return a DataProvider<Any> that completes when the update finishes (so that the caller knows when the language has, indeed, been updated).

Reply:
HI Ben getting this from here allows us to either write from updateAppLanguage or read from getAppLanguageSelection, please let me know if i am wrong on this.

Reply:
The problem is that the returned DataProvider will actually first return the previous selection state, then the latest after the update completes (due to Dataproviders having eventual consistency). This is an issue in this case because it makes the updateAppLanguage API less predictable. We really ought to return a one-off DataProvider here that only finishes when the update does (similar to how we do this in ProfileManagementController, though not necessarily that complicated). E.g. something like:

```kotlin
return dataProviders.createInMemoryDataProviderAsync(...newId...) { AsyncResult.Success(deferred.await()) }
```

The above provider will *only* finish once the deferred is completed, providing some predictability in the API.

Reply:
THanks, updated.